### DISCUSSION TITLE
Auto-Merge ApiKey

### DISCUSSION THREAD

**Discussion by Tsangares** [Tsangares/stoned_ape_tools#61] [Feature Requests]
As it says. The feature is made already. When I hit merge api-key please have it store in the plugin storage & re-populate upon refresh. 

### Issue
Must include a skip statement on all stars related to the owner.

**Comment by Skilly146** [comment]
Once the code is skipping all stars owned by the user do we even need caching? If we don't touch user owned stars it should still be showing the same data from the NP server. As long as the code works as intended and shows merged API data as well it should work perfectly

**Comment by anicolao** [comment]
I don't know why the merge API key code should need to skip the current player's stars. The data ought to be identical. NPA caches the API keys in the API keys report and merges all of them when the user presses the `(` hotkey, which I think is safer behaviour than merge on load in case there is some unknown bug. Caching the scan data itself is also worthwhile for speed reasons.

**Reply by Skilly146** [reply]
The problem is with displaying any changes the user makes after refresh. Take warpgates for example, if the UI is showing an allies view of a warpgated star owned by the user then when the user destroys that gate and refreshes the page the warpgate will still show up as being there since the allies view of the users stars hasn't updated yet. 

We would probably have similar problems with changing carrier orders and infrastructure changes, etc, but that is a bug we actually saw and is simpler to describe

**Comment by anicolao** [comment]
That bug means the merge is being done in the wrong order. The merge should start with an ally, do another ally, and finally merge your own scan data. There should be no need to skip stars if you merge your own scan data last. So far as I can see, NPA implements this correctly, so I don't know how you saw a bug.

**Reply by anicolao** [reply]
I stand corrected, NPA does not do this properly. Visible stars should only override invisible stars but visible stars currently do overwrite the user's existing scan data. It can be fixed with
```
        if (star.v !== "0" && universe.galaxy.stars[s].v === "0") {
```
